### ISSUE TITLE
No ability to query by custom mapping types

### ISSUE THREAD

**Issue by petr-buchin** [doctrine/mongodb-odm#2103]
### Bug Report

|    Q        |   A
|------------ | ------
| BC Break    | no
| Version     | 2.0.2

#### Summary

Currently ODM does not support querying by any PHP object values, except `DateTimeInterface` and `ObjectId`, even if there are custom mapping types to convert this values to their DB representation.
This makes using of ODM pretty weird, when you can use your value objects in your documents (entities), but you must query this documents by database values.

After a little research I found that problem is in the `Type::getTypeFromPHPVariable()` method, which looks unfinished:

```
if (is_object($variable)) {
      if ($variable instanceof DateTimeInterface) {
          return self::getType('date');
      }

      if ($variable instanceof ObjectId) {
          return self::getType('id');
      }
} else {
//...
}
```

Here ODM explicitly ignores any objects except `DateTimeInterface` and `ObjectId`.

I believe that maybe this behavior is a result of ODM not having an explicit way to map PHP value to custom mapping type. However, there is a mapping from field name to field type, and this info may be used to properly convert value.

#### Current behavior

No ability to search by PHP object values, used in custom mapping types.

#### How to reproduce

1. Create custom mapping types, which converts php object to some scalar value for storing in DB.
2. Create document, which uses this custom mapping type.
3. Try to query by a "php value" used in custom type.
4. Query will not find anything ever,

#### Expected behavior

I expect to be able to query by PHP objects, used as a type of a field in my document.

**Comment by alcaeus** [comment]
That is unfortunately known. I’ll put this on the roadmap, but can’t say when we’ll get to it.

**Comment by petr-buchin** [comment]
@alcaeus from what I see, this is pretty straightforward to add this feature. Actually, you can see my PR, which adds this feature at least for non-array values.

If you have any comments on that PR, I would be happy to fix it.

**Comment by alcaeus** [comment]
Fixed in #2104.
